core-backend/common/src/lib.rs

```diff
@@ -106,14 +107,21 @@ pub trait Environment<E: Ext + IntoExtInfo + 'static>: Sized {
     fn get_stack_mem_end(&mut self) -> Option<WasmPageNumber>;
 
     /// Returns host address of wasm memory buffer. Needed for lazy-pages
-    fn get_wasm_memory_begin_addr(&mut self) -> u64;
+    fn get_wasm_memory_begin_addr(&self) -> WasmBeginAddress;
 
     /// Run setuped instance starting at `entry_point` - wasm export function name.
-    /// - IMPORTANT: env is in inconsistent state after execution.
-    fn execute(&mut self, entry_point: &str) -> Result<BackendReport, BackendError>;
+    /// Also runs `post_execution_handler` after running instance at provided entry point.
+    fn execute<Pre, Post>(
+        self,
+        entry_point: &str,
+        pre_execution_handler: Option<Pre>,
+        post_execution_handler: Post,
```

Comment:
I would also set post_execution handler to Option

Reply:
discussed